File: serial_exercise/python/serial_test.py
import unittest
from serial import generate_serial_numbers


class TestSerial(unittest.TestCase):
    def test_empty_list(self):
        self.assertEqual(0, len(generate_serial_numbers('', 0)))
    
    def testSimpleSequentialList(self):
        serials = generate_serial_numbers('a2010', 10)
        self.assertTrue('a2010' in serials)
        self.assertTrue('a2011' in serials)
        self.assertTrue('a2012' in serials)
        self.assertTrue('a2013' in serials)
        self.assertTrue('a2014' in serials)
        self.assertTrue('a2015' in serials)
        self.assertTrue('a2016' in serials)
        self.assertTrue('a2017' in serials)
        self.assertTrue('a2018' in serials)
        self.assertTrue('a2019' in serials)
        self.assertEqual(10, len(serials))
    
    def testSimpleOffsetList(self):
        serials = generate_serial_numbers('a2010', 10, 2);
        self.assertTrue('a2010' in serials)
        self.assertTrue('a2012' in serials)
        self.assertTrue('a2014' in serials)
        self.assertTrue('a2016' in serials)
        self.assertTrue('a2018' in serials)
        self.assertTrue('a201a' in serials)
        self.assertTrue('a201c' in serials)
        self.assertTrue('a201e' in serials)
        self.assertTrue('a201g' in serials)
        self.assertTrue('a201i' in serials)
        self.assertEqual(10, len(serials))
    
    def testSmallNumber(self):
        serials = generate_serial_numbers('0', 11, 4);
        self.assertTrue('0' in serials)
        self.assertTrue('4' in serials)
        self.assertTrue('8' in serials)
        self.assertTrue('c' in serials)
        self.assertTrue('g' in serials)
        self.assertTrue('k' in serials)
        self.assertTrue('o' in serials)
        self.assertTrue('s' in serials)
        self.assertTrue('w' in serials)
        self.assertTrue('10' in serials)
        self.assertTrue('14' in serials)
        self.assertEqual(11, len(serials))
    
    def testBadCount(self):
        self.assertRaises(Exception, generate_serial_numbers, 'a2010', -1)
    
    def testBadOffset(self):
        self.assertRaises(Exception, generate_serial_numbers, 'a2010', 0, -1)

if __name__ == '__main__':
    unittest.main()

File: serial_exercise/php/SerialTest.php
<?php

include "Serial.php";

class SerialTest extends PHPUnit_Framework_TestCase
{
    public function testEmptyList()
    {
       $this->assertEquals(0, count(generateSerialNumbers('', 0, 0)));
    }
    
    /**
     * @expectedException Exception
     */
    public function testBadCount()
    {
        $serials = generateSerialNumbers('a2010', -1, 1);
        $this->assertEquals(0, count($serials));
    }
    
    /**
     * @expectedException Exception
     */
    public function testBadOffset()
    {
        $serials = generateSerialNumbers('a2010', 0,-1);
        $this->assertEquals(0, count($serials));
    }
    
    public function testSimpleSequentialList()
    {
        $serials = generateSerialNumbers('a2010', 10, 1);
        /*
        foreach($serials as $k){
            print($k."\n");
        }
        */
        $this->assertTrue(in_array('a2010', $serials));
        $this->assertTrue(in_array('a2011', $serials));
        $this->assertTrue(in_array('a2012', $serials));
        $this->assertTrue(in_array('a2013', $serials));
        $this->assertTrue(in_array('a2014', $serials));
        $this->assertTrue(in_array('a2015', $serials));
        $this->assertTrue(in_array('a2016', $serials));
        $this->assertTrue(in_array('a2017', $serials));
        $this->assertTrue(in_array('a2018', $serials));
        $this->assertTrue(in_array('a2019', $serials));
        $this->assertEquals(10, count($serials));
    }
    
    public function testSimpleOffsetList()
    {
        $serials = generateSerialNumbers('a2010', 10, 2);
        /*
        foreach($serials as $k){
            print($k."\n");
        }*/
        
        $this->assertTrue(in_array('a2010', $serials));
        $this->assertTrue(in_array('a2012', $serials));
        $this->assertTrue(in_array('a2014', $serials));
        $this->assertTrue(in_array('a2016', $serials));
        $this->assertTrue(in_array('a2018', $serials));
        $this->assertTrue(in_array('a201a', $serials));
        $this->assertTrue(in_array('a201c', $serials));
        $this->assertTrue(in_array('a201e', $serials));
        $this->assertTrue(in_array('a201g', $serials));
        $this->assertTrue(in_array('a201i', $serials));
        $this->assertEquals(10, count($serials));
    }
    
    public function testSmallNumberList()
    {
        $serials = generateSerialNumbers('0', 5, 2);
        /*
        foreach($serials as $k){
            print($k."\n");
        }*/
        
        $this->assertTrue(in_array('0', $serials));
        $this->assertTrue(in_array('2', $serials));
        $this->assertTrue(in_array('4', $serials));
        $this->assertTrue(in_array('6', $serials));
        $this->assertTrue(in_array('8', $serials));
        $this->assertEquals(5, count($serials));
    }
}